hack/generated/pkg/testcommon/test_context.go

```diff
@@ -6,47 +6,160 @@ Licensed under the MIT license.
+	"github.com/dnaeon/go-vcr/cassette"
+	"github.com/dnaeon/go-vcr/recorder"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
 )
 
+type TestContext struct {
+	AzureRegion  string
+	NameConfig   *ResourceNameConfig
+	RecordReplay bool
+}
+
+type PerTestContext struct {
+	TestContext
+	T                   *testing.T
+	AzureClientRecorder *recorder.Recorder
+	AzureClient         armclient.Applier
+	Namer               ResourceNamer
+	TestName            string
+}
+
 // If you modify this make sure to modify the cleanup-test-azure-resources target in the Makefile too
 const ResourcePrefix = "k8sinfratest"
-const DefaultTestRegion = "westus" // Could make this an env variable if we wanted
-
-type TestContext struct {
-	Namer       *ResourceNamer
-	AzureClient armclient.Applier
 
-	AzureRegion       string
-	AzureSubscription string
+func NewTestContext(region string, recordReplay bool) TestContext {
+	return TestContext{
+		AzureRegion:  region,
+		RecordReplay: recordReplay,
+		NameConfig:   NewResourceNameConfig(ResourcePrefix, "-", 6),
+	}
 }
 
-func NewTestContext(region string) (*TestContext, error) {
-	armClient, err := armclient.NewAzureTemplateClient(armclient.WithDefaultRetries())
+func (tc TestContext) ForTest(t *testing.T) (PerTestContext, error) {
+	authorizer, subscriptionID, recorder, err := createRecorder(t.Name(), tc.RecordReplay)
 	if err != nil {
-		return nil, errors.Wrapf(err, "creating armclient")
+		return PerTestContext{}, errors.Wrapf(err, "creating recorder")
 	}
 
-	subscription, ok := os.LookupEnv("AZURE_SUBSCRIPTION_ID")
-	if !ok {
-		return nil, errors.New("couldn't find AZURE_SUBSCRIPTION_ID")
+	armClient, err := armclient.NewAzureTemplateClient(authorizer, subscriptionID)
+	if err != nil {
+		return PerTestContext{}, errors.Wrapf(err, "creating ARM client")
 	}
 
-	return &TestContext{
-		AzureClient:       armClient,
-		AzureRegion:       region,
-		AzureSubscription: subscription,
-		Namer:             NewResourceNamer(ResourcePrefix, "-", 6),
+	// replace the ARM client transport (a bit hacky)
+	httpClient := armClient.RawClient.Sender.(*http.Client)
+	httpClient.Transport = recorder
+
+	t.Cleanup(func() {
+		log.Printf("stopping ARM client recorder")
+		err := recorder.Stop()
+		if err != nil {
+			// cleanup function should not error-out
+			log.Printf("unable to stop ARM client recorder: %s", err.Error())
+		}
+	})
+
+	return PerTestContext{
+		TestContext:         tc,
+		T:                   t,
+		Namer:               tc.NameConfig.NewResourceNamer(t.Name()),
+		AzureClient:         armClient,
+		AzureClientRecorder: recorder,
+		TestName:            t.Name(),
 	}, nil
 }
 
-func (tc *TestContext) NewTestResourceGroup() *resources.ResourceGroup {
+func createRecorder(testName string, recordReplay bool) (autorest.Authorizer, string, *recorder.Recorder, error) {
+	cassetteName := "fixtures/" + testName
+
+	var err error
+	var r *recorder.Recorder
+	if recordReplay {
```

Comment:
minor: slightly confusing variable name...
are we recording or are we replaying? Seems it would be cleaner as `replay` or `record`

Reply:
(once the `recorder` is created, it _does_ report which mode it is in (record/replay) and we use that below, but you can't start it in one of these modes)